File: prince-of-versions/src/main/java/co/infinum/princeofversions/CheckResult.java
package co.infinum.princeofversions;

import android.support.annotation.NonNull;

import java.util.Map;

import javax.annotation.Nullable;

/**
 * Intermediate result of update check. This result contains following data:
 * <ul>
 * <li>Update status (one of REQUIRED_UPDATE_NEEDED, NEW_UPDATE_AVAILABLE or NO_UPDATE_AVAILABLE)</li>
 * <li>Update version</li>
 * <li>Notification type (ONCE or ALWAYS)</li>
 * <li>Metadata</li>
 * </ul>
 */
final class CheckResult {

    private UpdateStatus status;

    private Integer updateVersion;

    @Nullable
    private NotificationType notificationType;

    private Map<String, String> metadata;

    private UpdateInfo info;

    private CheckResult(UpdateStatus status, Integer updateVersion, @Nullable NotificationType notificationType,
        Map<String, String> metadata, UpdateInfo updateInfo) {
        this.status = status;
        this.updateVersion = updateVersion;
        this.notificationType = notificationType;
        this.metadata = metadata;
        this.info = updateInfo;
    }

    static CheckResult mandatoryUpdate(Integer version, Map<String, String> metadata, UpdateInfo updateInfo) {
        return new CheckResult(UpdateStatus.REQUIRED_UPDATE_NEEDED, version, null, metadata, updateInfo);
    }

    static CheckResult optionalUpdate(Integer version, NotificationType notificationType, Map<String, String> metadata,
        UpdateInfo updateInfo) {
        return new CheckResult(UpdateStatus.NEW_UPDATE_AVAILABLE, version, notificationType, metadata, updateInfo);
    }

    static CheckResult noUpdate(Integer version, Map<String, String> metadata, UpdateInfo updateInfo) {
        return new CheckResult(UpdateStatus.NO_UPDATE_AVAILABLE, version, null, metadata, updateInfo);
    }

    boolean hasUpdate() {
        return UpdateStatus.REQUIRED_UPDATE_NEEDED.equals(status) || UpdateStatus.NEW_UPDATE_AVAILABLE.equals(status);
    }

    int getUpdateVersion() {
        return updateVersion;
    }

    boolean isOptional() {
        if (hasUpdate()) {
            return UpdateStatus.NEW_UPDATE_AVAILABLE.equals(status);
        } else {
            throw new UnsupportedOperationException("There is no update available.");
        }
    }

    public UpdateInfo getInfo() {
        return info;
    }

    @Nullable
    NotificationType getNotificationType() {
        if (isOptional()) {
            return notificationType;
        } else {
            throw new UnsupportedOperationException("There is no optional update available.");
        }
    }

    UpdateStatus status() {
        return status;
    }

    Map<String, String> metadata() {
        return metadata;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof CheckResult)) {
            return false;
        }

        CheckResult result = (CheckResult) o;

        if (status != result.status) {
            return false;
        }
        if (getUpdateVersion() != (result.getUpdateVersion())) {
            return false;
        }
        if (notificationType != result.notificationType) {
            return false;
        }
        return metadata.equals(result.metadata);
    }

    @Override
    public int hashCode() {
        int result = status.hashCode();
        result = 31 * result + (getUpdateVersion());
        result = 31 * result + (getNotificationType() != null ? getNotificationType().hashCode() : 0);
        result = 31 * result + (metadata.hashCode());
        return result;
    }

    @Override
    public String toString() {
        return this.status + " " + this.info + " " + this.metadata + " " + this.notificationType;
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/CheckResult.kt
package co.infinum.princeofversions

/**
 * Intermediate result of update check.
 *
 * This result contains:
 * - Update status ([UpdateStatus])
 * - Update version
 * - Notification type ([NotificationType])
 * - Metadata from the update configuration
 * - General update information ([UpdateInfo])
 */
@ConsistentCopyVisibility
data class CheckResult private constructor(
    val status: UpdateStatus,
    val updateVersion: Int,
    val notificationType: NotificationType?,
    val metadata: Map<String, String?>,
    val info: UpdateInfo
) {

    companion object {

        fun mandatoryUpdate(
            version: Int,
            metadata: Map<String, String?>,
            updateInfo: UpdateInfo
        ): CheckResult = CheckResult(
            status = UpdateStatus.REQUIRED_UPDATE_NEEDED,
            updateVersion = version,
            notificationType = null,
            metadata = metadata,
            info = updateInfo
        )

        fun optionalUpdate(
            version: Int,
            notificationType: NotificationType,
            metadata: Map<String, String?>,
            updateInfo: UpdateInfo
        ): CheckResult = CheckResult(
            status = UpdateStatus.NEW_UPDATE_AVAILABLE,
            updateVersion = version,
            notificationType = notificationType,
            metadata = metadata,
            info = updateInfo
        )

        fun noUpdate(
            version: Int,
            metadata: Map<String, String?>,
            updateInfo: UpdateInfo
        ): CheckResult = CheckResult(
            status = UpdateStatus.NO_UPDATE_AVAILABLE,
            updateVersion = version,
            notificationType = null,
            metadata = metadata,
            info = updateInfo
        )
    }

    /**
     * Checks if the result indicates that an update (either optional or mandatory) is available.
     * @return True if an update is available, false otherwise.
     */
    fun hasUpdate(): Boolean =
        status == UpdateStatus.REQUIRED_UPDATE_NEEDED || status == UpdateStatus.NEW_UPDATE_AVAILABLE

    /**
     * Checks if the available update is optional.
     * @return True if the update is optional.
     * @throws UnsupportedOperationException if no update is available.
     */
    fun isOptional(): Boolean =
        if (hasUpdate()) status == UpdateStatus.NEW_UPDATE_AVAILABLE
        else throw UnsupportedOperationException("There is no update available.")

    /**
     * Safely returns the notification type for an optional update.
     * @return The notification type if the update is optional.
     * @throws UnsupportedOperationException if the update is not optional.
     */
    fun safeNotificationType(): NotificationType? =
        if (isOptional()) notificationType
        else throw UnsupportedOperationException("There is no optional update available.")
}
